### Segmenting and Clustering Neighbourhoods in Toronto

_Author: R Forest_

__Part 1__

In [68]:
#Import necessary plugins

import pandas as pd
import requests
!pip install beautifulsoup4
# also need to install a parser for beautiful soup
!pip install lxml
from bs4 import BeautifulSoup

print("Import successful")

Import successful


_Scrap the list of postal codes from Wiki page_

In [69]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

In [70]:
soup = BeautifulSoup(source, 'xml')

In [71]:
table=soup.find('table')

In [72]:
# Creating dataframe using 3 headings; PostalCode, Borough and Neighborhood
column_names = ['Postalcode', 'Borough', 'Neighborhood']
df = pd.DataFrame(columns = column_names)

In [73]:
# Search all the postcode, borough, neighbourhood 
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [74]:
df.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

__Data Cleaning__

_Here we will remove the 'not assigned' fields of data_

In [75]:
df = df[df['Borough']!= 'Not assigned']

In [76]:
df.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

__Now will apply merging and grouping__

In [77]:
temp_df=df.groupby('Postalcode')['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
temp_df=temp_df.reset_index(drop=False)
temp_df.rename(columns={'Neighborhood':'Neighborhood_joined'},inplace=True)

In [78]:
df_merge = pd.merge(df, temp_df, on='Postalcode')

In [79]:
df_merge.drop(['Neighborhood'],axis=1,inplace=True)

In [80]:
df_merge.drop_duplicates(inplace=True)

In [81]:
df_merge.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)

In [82]:
df_merge.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [83]:
df_merge.shape

(103, 3)

_As you can see, the above shape is 103, 3_

## Part 2 

_Obtaining the co-ordinates for each location using GeoCoder_

In [84]:
import numpy as np

In [85]:
def get_geocode(postal_code):
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

In [86]:
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')

In [87]:
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [88]:
#Now we can merge the two tables together with the coordinate data
geo_df.rename(columns={'Postal Code':'Postalcode'},inplace=True)
geo_merged = pd.merge(geo_df, df_merge, on='Postalcode')

In [89]:
geo_data=geo_merged[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]

In [90]:
geo_data.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

_Above you can see the merged table_

## Part 3 

_We will now look at the data from above and use it to generate a map showing the clustering within Toronto._

In [91]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import os
from sklearn.cluster import KMeans
import folium 
import matplotlib.cm as cm
import matplotlib.colors as colors


print('Import success!')

Import success!


In [92]:
geo_data=geo_merged[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]
geo_data.head()

Postalcode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [93]:
toronto_data=geo_data[geo_data['Borough'].str.contains("Toronto")]
toronto_data.head()

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [94]:
CLIENT_ID = 'QEF4SFROUUVOQKNPAFNNRDUW4ACAWSYYG312LM3BKEDDKPIZ' # your Foursquare ID
CLIENT_SECRET = 'CCBFW1WDVTUYDLPFX3SR0QBB5R2UKRYOZ3F1JDJ3PLQMPDRM' # your Foursquare Secret
VERSION = '20180604'

In [95]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [96]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

_Above we see the list of establishments within our clustering maps._

In [97]:
#Now to view the top of our table
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [98]:
#We can now group the table by Neighbourhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              67   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
North Toronto West,  Lawrence Park                                     18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     35   
St. James Town                                                         87   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      18   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                         7   
Dufferin, Dovercourt Village       

In [99]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'] )
toronto_onehot.shape

(1633, 239)

In [104]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0.000000            0.000000      0.000000        0.000000   
1  0.000000            0.000000      0.000000        0.000000   
2  0.000000            0.000000      0.000000        0.000000   
3  0.066667            0.066667      0.066667        0.133333   
4  0.000000            0.000000      0.000000        0.000000   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0         0.000000          0.000000                  0.0           0.0  ...   
1         0.000000          0.000000                  0.0           0.0  ...   
2         0.000000          0.000000                  0.0           0.0  ...   
3         0.133333          0.133333                  0.0           0.0  ...   
4         0.000000          0.000000                  0.0           0.0  ...   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0    0.0            0.0                       0.017544   
1               0.0    0.0            0.0                       0.000000   
2               0.0    0.0            0.0                       0.000000   
3               0.0    0.0            0.0                       0.000000   
4               0.0    0.0            0.0                       0.015385   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  \
0               0.0                    0.0  0.000000        0.0   
1               0.0                    0.0  0.000000        0.0   
2               0.0                    0.0  0.000000        0.0   
3               0.0                    0.0  0.000000        0.0   
4               0.0                    0.0  0.015385        0.0   

   Women's Store  Yoga Studio  
0            0.0     0.000000  
1            0.0     0.000000  
2            0.0     0.055556  
3            0.0     0.000000  
4            0.0     0.015385  

[5 rows x 239 columns]

In [105]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [106]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market           Cheese Shop              Beer Bar   
1        Breakfast Spot           Coffee Shop                   Gym   
2                   Spa  Gym / Fitness Center         Garden Center   
3       Airport Service      Airport Terminal         Boat or Ferry   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant            Restaurant   
1               Stadium                   Bar          Intersection   
2                Garden  Fast Food Restaurant        Farmers Market   
3       Harbor / Marina              Boutique   Rental Car Location   
4       Bubble Tea Shop          Burger Joint   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café          Cocktail Bar                 Lounge  
1                Bakery    Italian Restaurant             Restaurant  
2    Light Rail Station            Comic Shop            Pizza Place  
3           Coffee Shop      Sculpture Garden                Airport  
4      Department Store           Salad Place       Ramen Restaurant

### Making clusters by neighbourhood

In [107]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [108]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031             0.0                 Trail                   Pub   
41 -79.352188             2.0      Greek Restaurant           Coffee Shop   
42 -79.315572             2.0                   Gym     Food & Drink Shop   
43 -79.340923             2.0                  Café           Coffee Shop   
44 -79.388790             4.0                  Park              Bus Line   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37     Health Food Store               Dog Run            Dessert Shop   
41    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
42        Sandwich Place          Liquor Store              Board Shop   
43                Bakery             Gastropub     American Restaurant   
44           Swim School          Dessert Shop    Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue  \
37    Dim Sum Restaurant                        Diner   
41            Restaurant                 Liquor Store   
42         Burrito Place                   Restaurant   
43               Brewery                  Yoga Studio   
44     Electronics Store  Eastern European Restaurant   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
37             Discount Store   Distribution Center            Yoga Studio  
41                  Bookstore                  Café                 Lounge  
42         Italian Restaurant  Fast Food Restaurant      Fish & Chips Shop  
43  Latin American Restaurant        Ice Cream Shop            Fish Market  
44        Dumpling Restaurant            Donut Shop       Doner Restaurant

In [109]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               2                                        Berczy Park   
1               2       Brockton, Parkdale Village, Exhibition Place   
2               2  Business reply mail Processing Centre, South C...   
3               2  CN Tower, King and Spadina, Railway Lands, Har...   
4               2                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop        Farmers Market           Cheese Shop   
1                  Café        Breakfast Spot           Coffee Shop   
2           Yoga Studio                   Spa  Gym / Fitness Center   
3        Airport Lounge       Airport Service      Airport Terminal   
4           Coffee Shop        Sandwich Place    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0              Beer Bar                Bakery    Seafood Restaurant   
1                   Gym               Stadium                   Bar   
2         Garden Center                Garden  Fast Food Restaurant   
3         Boat or Ferry       Harbor / Marina              Boutique   
4                  Café       Bubble Tea Shop          Burger Joint   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Restaurant                  Café          Cocktail Bar   
1          Intersection                Bakery    Italian Restaurant   
2        Farmers Market    Light Rail Station            Comic Shop   
3   Rental Car Location           Coffee Shop      Sculpture Garden   
4   Japanese Restaurant      Department Store           Salad Place   

  10th Most Common Venue  
0                 Lounge  
1             Restaurant  
2            Pizza Place  
3                Airport  
4       Ramen Restaurant

NameError: name 'Nominatim' is not defined

In [111]:
#Need to install an addition within geopy (Nominatim) - this is a street maps open source tool
!pip install geopy

In [112]:
from geopy.geocoders import Nominatim 

In [113]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.6534817, -79.3839347.


### Time to create the map!

In [126]:
#Having error in the map generation. Will try to re-import packages

import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import os
from sklearn.cluster import KMeans
import folium 
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm
import matplotlib.colors as colors


print('Libraries imported.')

Libraries imported.


In [128]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

##### Map may not be available on GitHub.